### DISCUSSION TITLE
Issue with CookieTokenStore delimiter

### DISCUSSION THREAD

**Discussion by adossant-redhat** [keycloak/keycloak#10025] [Ideas]
We have an issue with CookieTokenStore and the string it uses as a delimiter when it combines together the access token, ID token and refresh token strings to store in a cookie.
In our environment, the cookie parsing break every few hundred updates of the refresh token due to this.

Currently the delimiter is set to ”___” (three underscores). These tokens are all Base64Url encoded, which means that the underscore is a valid character in the encoded string. This can lead to that the signature part of the tokens can have ”___” in them, or have another underscore adjacent to the delimiter, breaking the parsing of the cookie.

We suggest that the delimiter string is changed to a character that is not a valid character in the Base64Url encoding but still adheres to RFC 6265 for allowed characters in a cookie value, for example !, + or &.
This seems to works for us (below a git diff), can you see any downside?

~~~
@@ -34,11 +34,12 @@ import org.keycloak.representations.IDToken;
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
 public class CookieTokenStore {
 
     private static final Logger log = Logger.getLogger(CookieTokenStore.class);
-    private static final String DELIM = "___";
+    private static final String DELIM = "&";
~~~

NOTE: Set the settings to reproduce: 
           - always-refresh-token=true 
           - token-store=cookie

**Comment by adossant-redhat** [comment]
References:
https://github.com/keycloak/keycloak/blob/master/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/CookieTokenStore.java
https://issues.redhat.com/browse/KEYCLOAK-1163
http://www.ietf.org/rfc/rfc6265.txt

**Reply by sguilhen** [reply]
@hmlnarik @pedroigor Whats your take on this?

**Reply by pedroigor** [reply]
IIRC, this is the second time we have a similar issue with the delimiter we use in cookies. See https://issues.redhat.com/browse/KEYCLOAK-1163.

TBH, I would just provide a configuration option and avoid this popping out every time someones find an issue.

We are planning to deprecate adapters too and I'm not sure if we are going to make changes like that.

@mposolda @stianst What are your thoughts?

**Reply by mposolda** [reply]
+1 for the configuration option for this with keep the defaults as is (as I am not 100% sure about other side-effects of changing this like backwards compatibility etc...)

IMO this will be community contribution

And as @pedroigor mentioned, it should be noted that adapters are going to be deprecated/removed (some of them already are).

**Comment by adossant-redhat** [comment]
@pedroigor Thank you very much for the feedback!
